File: examples/frc_2022_shooter/src/main.cpp
// Copyright (c) Sleipnir contributors

#include <cmath>
#include <numbers>
#include <print>

#include <Eigen/Core>
#include <sleipnir/optimization/problem.hpp>

// FRC 2022 shooter trajectory optimization.
//
// This program finds the initial velocity, pitch, and yaw for a game piece to
// hit the 2022 FRC game's target that minimizes either time-to-target or
// initial velocity (see minimize() calls below).
//
// This optimization problem formulation uses direct transcription of the flight
// dynamics, including air resistance.

using Eigen::Vector3d;
using Vector6d = Eigen::Vector<double, 6>;

constexpr double field_width = 8.2296;    // 27 ft -> m
constexpr double field_length = 16.4592;  // 54 ft -> m
constexpr Vector6d target_wrt_field{
    {field_length / 2.0}, {field_width / 2.0}, {2.64}, {0.0}, {0.0}, {0.0}};
[[maybe_unused]]
constexpr double target_radius = 0.61;               // m
constexpr double cone_angle = std::numbers::pi / 4;  // rad
constexpr Eigen::Vector3d g{{0.0}, {0.0}, {9.806}};  // m/s²

slp::VariableMatrix<double> f(const slp::VariableMatrix<double>& x) {
  using namespace slp::slicing;

  // x' = x'
  // y' = y'
  // z' = z'
  // x" = −F_D(v)/m v̂_x
  // y" = −F_D(v)/m v̂_y
  // z" = −g − F_D(v)/m v̂_z
  //
  // Per https://en.wikipedia.org/wiki/Drag_(physics)#The_drag_equation:
  //   F_D(v) = ½ρv²C_D A
  //   ρ is the fluid density in kg/m³
  //   v is the velocity magnitude in m/s
  //   C_D is the drag coefficient (dimensionless)
  //   A is the cross-sectional area of a circle in m²
  //   m is the mass in kg
  //   v̂ is the velocity direction unit vector
  constexpr double ρ = 1.204;  // kg/m³
  auto v = x[slp::Slice{3, 6}, _];
  slp::Variable v2 = v.T() * v;
  constexpr double C_D = 0.5;
  constexpr double r = 0.15;                      // m
  constexpr double A = std::numbers::pi * r * r;  // m²
  auto F_D = 0.5 * ρ * v2 * C_D * A;

  constexpr double m = 0.283;  // kg
  auto v_hat = v / sqrt(v2);
  return slp::block<double>({{v}, {-g - F_D / m * v_hat}});
}

#ifndef RUNNING_TESTS
int main() {
  // Robot initial state
  constexpr Vector6d robot_wrt_field{{field_length / 4.0},
                                     {field_width / 4.0},
                                     {0.0},
                                     {1.524},
                                     {-1.524},
                                     {0.0}};

  constexpr double max_initial_velocity = 10.0;  // m/s

  Vector6d shooter_wrt_robot{{0.0}, {0.0}, {1.2}, {0.0}, {0.0}, {0.0}};
  Vector6d shooter_wrt_field = robot_wrt_field + shooter_wrt_robot;

  slp::Problem<double> problem;

  // Set up duration decision variables
  constexpr int N = 50;
  auto T = problem.decision_variable();
  problem.subject_to(T >= 0);
  T.set_value(1);
  auto dt = T / N;

  // Ball state in field frame
  //
  //     [x position]
  //     [y position]
  //     [z position]
  // x = [x velocity]
  //     [y velocity]
  //     [z velocity]
  auto X = problem.decision_variable(6, N);

  auto p = X.block(0, 0, 3, N);
  auto p_x = X.row(0);
  auto p_y = X.row(1);
  auto p_z = X.row(2);

  auto v = X.block(3, 0, 3, N);
  auto v_x = X.row(3);
  auto v_y = X.row(4);
  auto v_z = X.row(5);

  auto v0_wrt_shooter =
      X.block(3, 0, 3, 1) - shooter_wrt_field.segment(3, 3).eval();

  // Position initial guess is linear interpolation between start and end
  // position
  for (int k = 0; k < N; ++k) {
    p_x[k].set_value(std::lerp(shooter_wrt_field(0), target_wrt_field(0),
                               static_cast<double>(k) / N));
    p_y[k].set_value(std::lerp(shooter_wrt_field(1), target_wrt_field(1),
                               static_cast<double>(k) / N));
    p_z[k].set_value(std::lerp(shooter_wrt_field(2), target_wrt_field(2),
                               static_cast<double>(k) / N));
  }

  // Velocity initial guess is max initial velocity toward target
  Vector3d uvec_shooter_to_target =
      (target_wrt_field.segment(0, 3) - shooter_wrt_field.segment(0, 3))
          .normalized();
  for (int k = 0; k < N; ++k) {
    v.col(k).set_value(robot_wrt_field.segment(3, 3) +
                       max_initial_velocity * uvec_shooter_to_target);
  }

  // Shooter initial position
  problem.subject_to(p.col(0) == shooter_wrt_field.block(0, 0, 3, 1));

  // Require initial velocity is below max
  //
  //   √(v_x² + v_y² + v_z²) ≤ vₘₐₓ
  //   v_x² + v_y² + v_z² ≤ vₘₐₓ²
  problem.subject_to(slp::pow(v_x[0] - robot_wrt_field[3], 2) +
                         slp::pow(v_y[0] - robot_wrt_field[4], 2) +
                         slp::pow(v_z[0] - robot_wrt_field[5], 2) <=
                     max_initial_velocity * max_initial_velocity);

  // Keep-out region (cylinder with conic bowl)
  //
  //            r_t
  //           |----|
  //   ___ ___  ____
  //    |   |  |   /
  //    |  h|  |  /
  // z_t|   |  |γ/    z
  //    |  ___ |/     ^
  //    |      |      |
  //   ___     |      ⋅-> x
  //
  // (x_c, y_c, z_c) is the tip of the cone, r_t is target radius, z_t is
  // target height.
  //
  //   tan(γ) = r_t/h
  //   h = r_t/tan(γ)
  //
  //   z_c = z_t - h
  //   z_c = z_t - r_t/tan(γ)
  //
  // Constrain ball to be either outside cylinder or inside cone.
  // First, get cylindrical coordinates with respect to cone origin.
  //
  //   r' = hypot(x - x_c, y - y_c)
  //   z' = z - z_c
  //
  // Cylinder:
  //   r' ≥ r_t
  //   (x - x_c)² + (y - y_c)² ≥ r_t²
  //   (x - x_c)² + (y - y_c)² - r_t² ≥ 0
  //
  // Cone:
  //   z' ≥ r'/tan(γ)
  //   z - z_c ≥ r'/tan(γ)
  //   (z - z_c) tan(γ) ≥ r'
  //   (z - z_c)² tan²(γ) ≥ (x - x_c)² + (y - y_c)²
  //   (z - z_c)² tan²(γ) - (x - x_c)² - (y - y_c)² ≥ 0
  //
  // Ball is in keep-out region if both constraints are violated.
  //
  //   max((x - x_c)² + (y - y_c)² - r_t²,
  //       (z - z_c)² tan²(γ) - (x - x_c)² - (y - y_c)²) ≥ 0
  auto& x_c = target_wrt_field[0, 0];
  auto& y_c = target_wrt_field[1, 0];
  auto z_c = target_wrt_field[2, 0] - target_radius / std::tan(cone_angle);
  for (int k = 0; k < N; ++k) {
    auto& x = p_x[k];
    auto& y = p_y[k];
    auto& z = p_z[k];

    auto x2 = slp::pow(x - x_c, 2);
    auto y2 = slp::pow(y - y_c, 2);
    auto z2 = slp::pow(z - z_c, 2);
    auto tan = std::tan(cone_angle);
    auto cylinder = x2 + y2 - target_radius * target_radius;
    auto cone = z2 * tan * tan - x2 - y2;

    problem.subject_to(slp::max(cylinder, cone) >= 0);
  }

  // Dynamics constraints - RK4 integration
  auto h = dt;
  for (int k = 0; k < N - 1; ++k) {
    auto x_k = X.col(k);
    auto x_k1 = X.col(k + 1);

    auto k1 = f(x_k);
    auto k2 = f(x_k + h / 2 * k1);
    auto k3 = f(x_k + h / 2 * k2);
    auto k4 = f(x_k + h * k3);
    problem.subject_to(x_k1 == x_k + h / 6 * (k1 + 2 * k2 + 2 * k3 + k4));
  }

  // Require final position is in center of target circle
  problem.subject_to(p.col(N - 1) == target_wrt_field.block(0, 0, 3, 1));

  // Require the final velocity is down
  problem.subject_to(v_z[N - 1] < 0.0);

  // Minimize time-to-target
  // problem.minimize(T);

  // Minimize initial velocity
  problem.minimize(v0_wrt_shooter.T() * v0_wrt_shooter);

  problem.solve({.diagnostics = true});

  // Initial velocity vector with respect to shooter
  Eigen::Vector3d v0 = v0_wrt_shooter.value();

  double velocity = v0.norm();
  std::println("Velocity = {:.03} ms", velocity);

  double pitch = std::atan2(v0[2], std::hypot(v0[0], v0[1]));
  std::println("Pitch = {:.03}°", pitch * 180.0 / std::numbers::pi);

  double yaw = std::atan2(v0[1], v0[0]);
  std::println("Yaw = {:.03}°", yaw * 180.0 / std::numbers::pi);

  std::println("Total time = {:.03} s", T.value());
  std::println("dt = {:.03} ms", dt.value() * 1e3);
}
#endif

File: examples/frc_2022_shooter/main.py
#!/usr/bin/env python3

"""
FRC 2022 shooter trajectory optimization.

This program finds the initial velocity, pitch, and yaw for a game piece to hit
the 2022 FRC game's target that minimizes either time-to-target or initial
velocity (see minimize() calls below).

This optimization problem formulation uses direct transcription of the flight
dynamics, including air resistance.
"""

import math

import matplotlib.pyplot as plt
import numpy as np
from numpy.linalg import norm
from sleipnir.autodiff import VariableMatrix
from sleipnir.optimization import Problem

field_width = 8.2296  # 27 ft -> m
field_length = 16.4592  # 54 ft -> m
target_wrt_field = np.array(
    [[field_length / 2.0], [field_width / 2.0], [2.64], [0.0], [0.0], [0.0]]
)
target_radius = 0.61  # m
g = 9.806  # m/s²


def lerp(a, b, t):
    return a + t * (b - a)


def f(x):
    # x' = x'
    # y' = y'
    # z' = z'
    # x" = −a_D(v_x)
    # y" = −a_D(v_y)
    # z" = −g − a_D(v_z)
    #
    # where a_D(v) = ½ρv² C_D A / m
    # (see https://en.wikipedia.org/wiki/Drag_(physics)#The_drag_equation)
    rho = 1.204  # kg/m³
    C_D = 0.5
    A = math.pi * 0.3
    m = 2.0  # kg
    a_D = lambda v: 0.5 * rho * v**2 * C_D * A / m

    v_x = x[3, 0]
    v_y = x[4, 0]
    v_z = x[5, 0]
    return VariableMatrix(
        [[v_x], [v_y], [v_z], [-a_D(v_x)], [-a_D(v_y)], [-g - a_D(v_z)]]
    )


def main():
    # Robot initial state
    robot_wrt_field = np.array(
        [[field_length / 4.0], [field_width / 4.0], [0.0], [1.524], [-1.524], [0.0]]
    )

    max_initial_velocity = 10.0  # m/s

    shooter_wrt_robot = np.array([[0.0], [0.0], [1.2], [0.0], [0.0], [0.0]])
    shooter_wrt_field = robot_wrt_field + shooter_wrt_robot

    problem = Problem()

    # Set up duration decision variables
    N = 10
    T = problem.decision_variable()
    problem.subject_to(T >= 0)
    T.set_value(1)
    dt = T / N

    # Ball state in field frame
    #
    #     [x position]
    #     [y position]
    #     [z position]
    # x = [x velocity]
    #     [y velocity]
    #     [z velocity]
    X = problem.decision_variable(6, N)

    p = X[:3, :]
    p_x = X[0, :]
    p_y = X[1, :]
    p_z = X[2, :]

    v = X[3:, :]
    v_x = X[3, :]
    v_y = X[4, :]
    v_z = X[5, :]

    v0_wrt_shooter = X[3:, :1] - shooter_wrt_field[3:, :]

    # Position initial guess is linear interpolation between start and end position
    for k in range(N):
        p_x[k].set_value(lerp(shooter_wrt_field[0, 0], target_wrt_field[0, 0], k / N))
        p_y[k].set_value(lerp(shooter_wrt_field[1, 0], target_wrt_field[1, 0], k / N))
        p_z[k].set_value(lerp(shooter_wrt_field[2, 0], target_wrt_field[2, 0], k / N))

    # Velocity initial guess is max initial velocity toward target
    uvec_shooter_to_target = target_wrt_field[:3, :] - shooter_wrt_field[:3, :]
    uvec_shooter_to_target /= norm(uvec_shooter_to_target)
    for k in range(N):
        v[:, k].set_value(
            robot_wrt_field[3:, :] + max_initial_velocity * uvec_shooter_to_target
        )

    # Shooter initial position
    problem.subject_to(p[:, :1] == shooter_wrt_field[:3, :])

    # Require initial velocity is below max
    #
    #   √(v_x² + v_y² + v_z²) ≤ vₘₐₓ
    #   v_x² + v_y² + v_z² ≤ vₘₐₓ²
    problem.subject_to(
        (v_x[0] - robot_wrt_field[3, 0]) ** 2
        + (v_y[0] - robot_wrt_field[4, 0]) ** 2
        + (v_z[0] - robot_wrt_field[5, 0]) ** 2
        <= max_initial_velocity**2
    )

    # Dynamics constraints - RK4 integration
    h = dt
    for k in range(N - 1):
        x_k = X[:, k]
        x_k1 = X[:, k + 1]

        k1 = f(x_k)
        k2 = f(x_k + h / 2 * k1)
        k3 = f(x_k + h / 2 * k2)
        k4 = f(x_k + h * k3)
        problem.subject_to(x_k1 == x_k + h / 6 * (k1 + 2 * k2 + 2 * k3 + k4))

    # Require final position is in center of target circle
    problem.subject_to(p[:, -1] == target_wrt_field[:3, :])

    # Require the final velocity is down
    problem.subject_to(v_z[-1] < 0.0)

    # Minimize time-to-target
    problem.minimize(T)

    # Minimize initial velocity
    # problem.minimize(v0_wrt_shooter.T @ v0_wrt_shooter)

    problem.solve(diagnostics=True)

    # Initial velocity vector with respect to shooter
    v0 = v0_wrt_shooter.value()

    velocity = norm(v0)
    print(f"Velocity = {velocity:.03f} m/s")

    pitch = math.atan2(v0[2, 0], math.hypot(v0[0, 0], v0[1, 0]))
    print(f"Pitch = {np.rad2deg(pitch):.03f}°")

    yaw = math.atan2(v0[1, 0], v0[0, 0])
    print(f"Yaw = {np.rad2deg(yaw):.03f}°")

    print(f"Total time = {T.value():.03f} s")
    print(f"dt = {dt.value() * 1e3:.03f} ms")

    plt.figure()
    ax = plt.axes(projection="3d")

    def plot_wireframe(ax, f, x_range, y_range, color):
        x, y = np.mgrid[x_range[0] : x_range[1] : 25j, y_range[0] : y_range[1] : 25j]

        # Need an (N, 2) array of (x, y) pairs.
        xy = np.column_stack([x.flat, y.flat])

        z = np.zeros(xy.shape[0])
        for i, pair in enumerate(xy):
            z[i] = f(pair[0], pair[1])
        z = z.reshape(x.shape)

        ax.plot_wireframe(x, y, z, color=color)

    # Ground
    plot_wireframe(ax, lambda x, y: 0.0, [0, field_length], [0, field_width], "grey")

    # Target
    ax.plot(
        target_wrt_field[0, 0],
        target_wrt_field[1, 0],
        target_wrt_field[2, 0],
        color="black",
        marker="x",
    )
    xs = []
    ys = []
    zs = []
    for angle in np.arange(0.0, 2.0 * math.pi, 0.1):
        xs.append(target_wrt_field[0, 0] + target_radius * math.cos(angle))
        ys.append(target_wrt_field[1, 0] + target_radius * math.sin(angle))
        zs.append(target_wrt_field[2, 0])
    ax.plot(xs, ys, zs, color="black")

    # Trajectory
    trajectory_x = p_x.value()[0, :]
    trajectory_y = p_y.value()[0, :]
    trajectory_z = p_z.value()[0, :]
    ax.plot(trajectory_x, trajectory_y, trajectory_z, color="orange")

    ax.set_box_aspect((field_length, field_width, np.max(trajectory_z)))

    ax.set_xlabel("X position (m)")
    ax.set_ylabel("Y position (m)")
    ax.set_zlabel("Z position (m)")

    plt.show()


if __name__ == "__main__":
    main()
